documentation/proposals/Proposal - Silk.NET 3.X Generic Maths.md

```diff
@@ -0,0 +1,884 @@
+    public string ToString([StringSyntax(StringSyntaxAttribute.NumericFormat)] string? format);
+    /// <summary>Returns the string representation of the current instance using the specified format string to format individual elements and the specified format provider to define culture-specific formatting.</summary>
+    /// <param name="format">A standard or custom numeric format string that defines the format of individual elements.</param>
+    /// <param name="formatProvider">A format provider that supplies culture-specific formatting information.</param>
+    /// <returns>The string representation of the current instance.</returns>
+    /// <remarks>This method returns a string in which each element of the vector is formatted using <paramref name="format" /> and <paramref name="formatProvider" />. The "&lt;" and "&gt;" characters are used to begin and end the string, and the format provider's <see cref="NumberFormatInfo.NumberGroupSeparator" /> property followed by a space is used to separate each element.</remarks>
+    /// <related type="Article" href="/dotnet/standard/base-types/standard-numeric-format-strings">Standard Numeric Format Strings</related>
+    /// <related type="Article" href="/dotnet/standard/base-types/custom-numeric-format-strings">Custom Numeric Format Strings</related>
+    public string ToString([StringSyntax(StringSyntaxAttribute.NumericFormat)] string? format, IFormatProvider? formatProvider);
+    public bool TryFormat(Span<char> destination, out int charsWritten, ReadOnlySpan<char> format, IFormatProvider? provider);
+    #endregion
+
+    #region Casts
+    public Vector2D<TOther> As<TOther>() where TOther : INumberBase<TOther>;
+    private Vector2D<TOther> AsChecked<TOther>() where TOther : INumberBase<TOther>;
+    public static explicit operator Vector2D<byte>(Vector2D<T> self);
+    public static explicit operator Vector2D<sbyte>(Vector2D<T> self);
+    public static explicit operator Vector2D<short>(Vector2D<T> self);
+    public static explicit operator Vector2D<ushort>(Vector2D<T> self);
+    public static explicit operator Vector2D<int>(Vector2D<T> self);
+    public static explicit operator Vector2D<uint>(Vector2D<T> self);
+    public static explicit operator Vector2D<long>(Vector2D<T> self);
+    public static explicit operator Vector2D<ulong>(Vector2D<T> self);
+    public static explicit operator Vector2D<Int128>(Vector2D<T> self);
+    public static explicit operator Vector2D<UInt128>(Vector2D<T> self);
+    public static explicit operator Vector2D<Half>(Vector2D<T> self);
+    public static explicit operator Vector2D<float>(Vector2D<T> self);
+    public static explicit operator Vector2D<double>(Vector2D<T> self);
+    public static explicit operator Vector2D<decimal>(Vector2D<T> self);
+    public static explicit operator Vector2D<Complex>(Vector2D<T> self);
+    public static explicit operator Vector2D<BigInteger>(Vector2D<T> self);
```

Comment:
The explicit operators are a case we aren't exposing for the new `Vector2<T>` and other types we'd define in the BCL

Notably the explicit conversions on `Vector<T>` do a reinterpret cast. This was arguably bad design, but it also means you may set the "wrong" expectations for users here.

Rather instead, we would expose them as `To<TOther>()` (or `ConvertTo*()`) methods if they actually convert or `As<TOther>()` methods if they reinterpret cast. This helps with readability and makes it clear there is more complex considerations involved in the conversion process.

Reply:
Matching the implementation for Silk.NET.Maths 2.X here. I also strongly believe a regular cast rather than a reinterpret cast is ideal, especially for mismatching sizes, and this aligns with behavior that users expect - no functions on this type do a reinterpret cast to a vector of different scalar, so having it differ here would be a larger concern.